source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc

```diff
@@ -127,19 +145,34 @@ void SimpleHttpCache::updateHeaders(const LookupContext& lookup_context,
     return;
   }
 
-  // https://www.rfc-editor.org/rfc/pdfrfc/rfc7234.txt.pdf
-  // 4.3.4 Freshening Stored Responses upon Validation
-  // use other header fields provided in the 304 (Not Modified)
-  // response to replace all instances of the corresponding header
-  // fields in the stored response.
-  //
   // Assumptions:
   // 1. The internet is fast, i.e. we get the result as soon as the server sends it.
-  // Race conditions would not be possible because we are always processing up-to-date data.
+  //    Race conditions would not be possible because we are always processing up-to-date data.
   // 2. No key collision for etag. Therefore, if etag matches it's the same resource.
   // 3. Backend is correct. etag is being used as a unique identifier to the resource
-  // TODO(tangsaidi) merge the header map instead of replacing it according to rfc7234
-  entry.response_headers_ = Http::createHeaderMap<Http::ResponseHeaderMapImpl>(response_headers);
+
+  // use other header fields provided in the new response to replace all instances
+  // of the corresponding header fields in the stored response
+  absl::flat_hash_set<Http::LowerCaseString> updatedHeaderFields;
+  response_headers.iterate(
+      [&entry, &updatedHeaderFields](
+          const Http::HeaderEntry& incoming_response_header) -> Http::HeaderMap::Iterate {
+        Http::LowerCaseString lower_case_key{incoming_response_header.key().getStringView()};
+        absl::string_view incoming_value{incoming_response_header.value().getStringView()};
+        if (headersNotToUpdate().contains(lower_case_key)) {
+          return Http::HeaderMap::Iterate::Continue;
+        }
+        // A single `setCopy` cannot be used here, because if there's a repeated header in the
```

Comment:
Thanks for changing to the simple setCopy call. I didn't realize that the impl itself does 2 lookups for this one call; seems not necessary but that's a topic for a different PR.

Can you update this comment now?

Reply:
Done :D